aptos-move/move-examples/fungible_asset/sources/fungible_asset.move

```diff
@@ -0,0 +1,389 @@
+    public fun transfer<T: key>(
+        fungible_asset_owner: &signer,
+        asset: &Object<T>,
+        amount: u64,
+        to: address
+    ) acquires FungibleAssetStore, FungibleAsset, FungibleAssetProperty {
+        // This ensures amount > 0;
+        let fa = withdraw(fungible_asset_owner, asset, amount);
+        deposit(fa, to);
+    }
+
+    public(friend) fun mint(
+        asset_addr: address,
+        amount: u64,
+    ): FungibleAsset {
+        assert!(amount != 0, error::invalid_argument(EAMOUNT_CANNOT_BE_ZERO));
+        FungibleAsset {
+            asset_addr,
+            balance: amount
+        }
+    }
+
+    public(friend) fun set_frozen_flag<T: key>(
+        fungible_asset_owner: address,
+        asset: &Object<T>,
+        frozen: bool
+    ) acquires FungibleAssetStore, FungibleAssetProperty {
+        let fungible_asset_obj_opt = get_fungible_asset_object(fungible_asset_owner, object_address(asset), true);
+        let fungible_asset_obj = option::destroy_some(fungible_asset_obj_opt);
+        borrow_global_mut<FungibleAssetProperty>(verify(&fungible_asset_obj)).frozen = frozen;
+    }
+
+    public(friend) fun burn(fungible_asset: FungibleAsset) {
+        let FungibleAsset {
+            asset_addr: _,
+            balance: _,
+        } = fungible_asset;
+    }
+
+    public(friend) fun withdraw_internal<T: key>(
+        account: address,
+        asset: &Object<T>,
+        amount: u64
+    ): FungibleAsset acquires FungibleAssetStore, FungibleAsset, FungibleAssetProperty {
+        let asset_address = object_address(asset);
+        let (stored_fa, property) = borrow_fungible_asset_mut_and_property(
+            account,
+            asset_address,
+            false /* create token object if not exist */
+        );
+
+        assert!(!property.frozen, error::invalid_argument(ESTILL_FROZEN));
+        let fungible_asset = withdraw_fungible_asset(stored_fa, amount);
+        // Clean up token obj if amount drops to 0 and not frozen (verified).
+        if (stored_fa.balance == 0) {
+            let fungible_asset_obj = remove_fungible_asset_object(account, asset_address);
+            let fungible_asset_obj_addr = verify(&fungible_asset_obj);
+            let FungibleAsset {
+                asset_addr,
+                balance: amount,
+            } = move_from<FungibleAsset>(fungible_asset_obj_addr);
+            let FungibleAssetProperty {
+                frozen,
+                delete_ref
+            } = move_from<FungibleAssetProperty>(fungible_asset_obj_addr);
+            assert!(asset_addr == asset_address, error::internal(EASSET_ADDRESS_MISMATCH));
+            assert!(amount == 0, error::internal(EBALANCE_NOT_ZERO));
+            assert!(!frozen, error::internal(ESTILL_FROZEN));
+            object::delete(delete_ref);
+        };
+        fungible_asset
+    }
+
+    /// Ensure the coin store exists. If not, create it.
+    fun ensure_fungible_asset_store(account_address: address) {
```

Comment:
probably a good inline function and could even make a borrow / borrow_mut inline